From 83f3f280eaadcbadcaaa91766708577631a3da6f Mon Sep 17 00:00:00 2001
From: Cheick TRAORE <cheick.traore-ext@st.com>
Date: Fri, 21 Jun 2024 12:04:06 +0200
Subject: [PATCH 0983/1695] drivers: stm32_uart: configure UART baudrate
 register

Add configuration of UART baudrate register in a UART driver probe. The
configuration of the baudrate register is done only for the UART device
used as serial console and once the UART is probed it is registered as
the serial console to use.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2425a1a5790b290266ed51b358e22e6b895673d4 drivers: stm32_uart: configure UART baudrate register]
Change-Id: Iccfd1d7ed5bccb072dd0ba549a5d25dd5b84de82
Signed-off-by: Cheick TRAORE <cheick.traore-ext@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/374893
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_uart.c | 197 ++++++++++++++++++++++++++++++++------
 1 file changed, 168 insertions(+), 29 deletions(-)

diff --git a/core/drivers/stm32_uart.c b/core/drivers/stm32_uart.c
index b1e8512bc..421f81680 100644
--- a/core/drivers/stm32_uart.c
+++ b/core/drivers/stm32_uart.c
@@ -46,6 +46,25 @@
 #define USART_ISR_TXE_TXFNF		BIT(7)
 #define USART_ISR_TXFE			BIT(23)
 
+/*
+ * UART Configuration register 1
+ *
+ * Bit 0 UE: UART Enable
+ * Bit 2 RE: Reception Enable
+ * Bit 3 TE: Transmission Enable
+ * Bit 15 OVER8: Oversampling by 8
+ * Bit 29 FIFOEN: UART Fifo Enable
+ */
+#define UART_REG_CR1_UE			BIT(0)
+#define UART_REG_CR1_RE			BIT(2)
+#define UART_REG_CR1_TE			BIT(3)
+#define UART_REG_CR1_OVER8		BIT(15)
+#define UART_REG_CR1_FIFOEN		BIT(29)
+
+#define UART_MAX_PRESC			256
+#define UART_REG_BRR_M_SHIFT		4
+#define UART_REG_BRR_MASK		GENMASK_32(15, 0)
+
 static vaddr_t loc_chip_to_base(struct serial_chip *chip)
 {
 	struct stm32_uart_pdata *pd = NULL;
@@ -138,59 +157,179 @@ static void register_non_secure_uart(struct stm32_uart_pdata *pd)
 
 }
 
-struct stm32_uart_pdata *stm32_uart_init_from_dt_node(void *fdt, int node)
+static TEE_Result stm32_uart_update_baudrate(struct stm32_uart_pdata *pd,
+					     char *params)
+{
+	uint64_t clk_rate = 0;
+	uint32_t brr = 0;
+	uint32_t baudrate = 0;
+	uint32_t mantissa = 0;
+	uint32_t uartdiv = 0;
+	uint32_t fraction = 0;
+	unsigned int i = 0;
+	int oversampling = 0;
+	uint32_t over8 = 0;
+	vaddr_t uart_base = io_pa_or_va(&pd->base, 1);
+	uint32_t presc[] = { 1, 2, 6, 8, 10, 12, 16, 32, 64, 128, 256 };
+
+	clk_rate = clk_get_rate(pd->clock);
+	if (!clk_rate || clk_enable(pd->clock))
+		panic();
+
+	baudrate = strtoul(params, NULL, 10);
+
+	for (i = 0; i < ARRAY_SIZE(presc); i++) {
+		uartdiv = UDIV_ROUND_NEAREST(clk_rate, baudrate * presc[i]);
+		if (uartdiv < 16) {
+			oversampling = 8;
+			uartdiv = uartdiv * 2;
+			over8 = UART_REG_CR1_OVER8;
+		} else {
+			oversampling = 16;
+		}
+		mantissa = SHIFT_U32(uartdiv / oversampling,
+				     UART_REG_BRR_M_SHIFT);
+		fraction = uartdiv % oversampling;
+		brr = mantissa | fraction;
+		if (brr <= UART_REG_BRR_MASK)
+			break;
+		if (i == ARRAY_SIZE(presc))
+			return TEE_ERROR_GENERIC;
+	}
+	io_write32(uart_base + UART_REG_BRR, brr);
+	io_write32(uart_base + UART_REG_PRESC, presc[i]);
+	io_write32(uart_base + UART_REG_CR1, UART_REG_CR1_UE |
+			   UART_REG_CR1_RE | UART_REG_CR1_TE |
+			   UART_REG_CR1_FIFOEN | over8);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result parse_dt(void *fdt, int node, struct stm32_uart_pdata *pd)
 {
-	TEE_Result res = TEE_ERROR_GENERIC;
-	struct stm32_uart_pdata *pd = NULL;
 	struct dt_node_info info = { };
+	TEE_Result res = TEE_ERROR_GENERIC;
 
-	fdt_fill_device_info(fdt, &info, node);
+	res = clk_dt_get_by_index(fdt, node, 0, &pd->clock);
+	if (res)
+		return res;
 
-	if (info.status == DT_STATUS_DISABLED)
-		return NULL;
+	res = pinctrl_get_state_by_name(fdt, node, "default", &pd->pinctrl);
+	if (res)
+		return res;
+	res = pinctrl_get_state_by_name(fdt, node, "sleep", &pd->pinctrl_sleep);
+	if (res)
+		return res;
+
+	fdt_fill_device_info(fdt, &info, node);
 
 	assert(info.reg != DT_INFO_INVALID_REG &&
 	       info.reg_size != DT_INFO_INVALID_REG_SIZE);
 
-	pd = calloc(1, sizeof(*pd));
-	if (!pd)
-		panic();
-
 	pd->chip.ops = &stm32_uart_serial_ops;
 	pd->base.pa = info.reg;
 	pd->secure = (info.status == DT_STATUS_OK_SEC);
 
-	res = clk_dt_get_by_index(fdt, node, 0, &pd->clock);
-	if (res) {
-		EMSG("Failed to get clock: %#"PRIx32, res);
-		panic();
-	}
-
-	res = clk_enable(pd->clock);
-	if (res)
-		panic();
-
 	assert(cpu_mmu_enabled());
 	pd->base.va = (vaddr_t)phys_to_virt(pd->base.pa,
-					    pd->secure ? MEM_AREA_IO_SEC :
-					    MEM_AREA_IO_NSEC, info.reg_size);
+					pd->secure ? MEM_AREA_IO_SEC :
+					MEM_AREA_IO_NSEC, info.reg_size);
 
-	res = pinctrl_get_state_by_name(fdt, node, "default", &pd->pinctrl);
-	if (res)
-		panic();
+	return TEE_SUCCESS;
+}
 
-	res = pinctrl_get_state_by_name(fdt, node, "sleep", &pd->pinctrl_sleep);
-	if (res && res != TEE_ERROR_ITEM_NOT_FOUND)
+static void setup_resources(struct stm32_uart_pdata *pd)
+{
+	if (clk_enable(pd->clock))
 		panic();
 
-	res = pinctrl_apply_state(pd->pinctrl);
-	if (res)
+	if (pinctrl_apply_state(pd->pinctrl))
 		panic();
 
 	if (pd->secure)
 		register_secure_uart(pd);
 	else
 		register_non_secure_uart(pd);
+}
+
+struct stm32_uart_pdata *stm32_uart_init_from_dt_node(void *fdt, int node)
+{
+	struct stm32_uart_pdata *pd = NULL;
+
+	if (fdt_get_status(fdt, node) == DT_STATUS_DISABLED)
+		return NULL;
+
+	pd = calloc(1, sizeof(*pd));
+	if (!pd)
+		panic();
+
+	if (parse_dt(fdt, node, pd)) {
+		/* stm32_uart_init_from_dt_node() expects resources are ready */
+		panic();
+	}
+
+	setup_resources(pd);
 
 	return pd;
 }
+
+static bool uart_is_for_console(void *fdt, int node, char **params)
+{
+	static int uart_console_node = -1;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (uart_console_node < 0) {
+		res = get_console_node_from_dt(fdt, &uart_console_node, NULL,
+					       params);
+		if (res == TEE_ERROR_ITEM_NOT_FOUND)
+			return false;
+		if (res != TEE_SUCCESS) {
+			EMSG("Error finding a console UART device");
+			return false;
+		}
+	}
+
+	return uart_console_node == node;
+}
+
+static TEE_Result stm32_uart_probe(const void *fdt, int node,
+				   const void *compt_data __unused)
+{
+	struct stm32_uart_pdata *pd = NULL;
+	static char *params;
+	TEE_Result res = TEE_SUCCESS;
+
+	pd = calloc(1, sizeof(*pd));
+	if (!pd)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = parse_dt((void *)fdt, node, pd);
+	if (res)
+		goto out;
+
+	setup_resources(pd);
+
+	if (uart_is_for_console((void *)fdt, node, &params) &&
+	    stm32_uart_update_baudrate(pd, params) == TEE_SUCCESS) {
+		register_serial_console(&pd->chip);
+		IMSG("UART console (%ssecure)", pd->secure ? "" : "non-");
+	}
+
+out:
+	if (res)
+		free(pd);
+
+	return res;
+}
+
+static const struct dt_device_match stm32_uart_match_table[] = {
+	{ .compatible = "st,stm32h7-uart" },
+	{ }
+};
+
+DEFINE_DT_DRIVER(stm32_uart_dt_driver) = {
+	.name = "stm32_uart",
+	.type = DT_DRIVER_UART,
+	.match_table = stm32_uart_match_table,
+	.probe = &stm32_uart_probe,
+};
-- 
2.47.2

